### ISSUE TITLE
Sample rate is fixed to 48000Hz

### ISSUE THREAD

**Issue by iamalexei** [Taiko2k/TauonMusicBox#1093]
Tauon version: 7.7.1
Platform: Flatpak
Tauon settings: "Avoid Resampling" is ON

Audio server: Pipewire 1.0.3
pipewire.conf: default.clock.allowed-rates = [ 44100 48000 ]

Tauon doesn't dynamically change sample rate depending on an audio file played despite reporting that it does:

```
$ flatpak run com.github.taiko2k.tauonmb
    ...
ph: Connected using samplerate 48000hz
ph: The samplerate changed, rewinding
ph: Connected using samplerate 44100hz
```

```
$ pw-cli ls
    ...
id 70, type PipeWire:Interface:Node/3
            object.serial = "654"
            factory.id = "6"
            client.id = "73"
            client.api = "pipewire-pulse"
            application.name = "tauon.py"
            node.name = "tauon.py"
            media.class = "Stream/Output/Audio"
            media.category = "Manager"
    ...
```

- File 1 (44100hz) played:

```
$ pw-cli i 70 | grep node.rate
*               node.rate = "1/48000"
```

- File 2 (48000hz) played:

```
$ pw-cli i 70 | grep node.rate
*               node.rate = "1/48000"
```

I've tried to play the same files with pw-play and mpv (--ao=pipewire and --ao=pulse) to check if something is wrong with my system/configuration, luckily it's not the case: pw-cli reports correct node.rate for each file.

**Comment by Taiko2k** [comment]
Ah not this again. Unfortunately Ive already spent too much time trying to fix this bug than makes sense. I thought it was fixed but yeah its kind of a known issue.

**Comment by iamalexei** [comment]
@Taiko2k Such a stubborn backend. What advantages it has for you compared to alternatives like libmpv or libVLC?

**Comment by Taiko2k** [comment]
The related library here is miniaudio. I was hoping that the developers of that library would impliment a native pipewire backend by now as opposed to using puleaudio that would have hopefully solved such issues. Otherwise the library is a good fit for Tauon. I don't think changing library now would be worth the development effort though.

**Comment by iamalexei** [comment]
@Taiko2k In general, I don't mind using PulseAudio output plugins with PipeWire. They are capable to change sample rate of the node per audio track, but the miniaudio's one is the first exception I've stumbled upon.

> I don't think changing library now would be worth the development effort though.

I hope you're right and it won't turn out to be the sunk cost fallacy.

**Comment by Taiko2k** [comment]
Actually the version of miniaudio currently used is a few versions out of date, its possible a fix exists in a newer version. Ive built a new flatpak with the most up to date version, if you'd like to test it. 

You can install this test flatpak using this command: (Uninstall the stable version first. When you want to switch back to stable, uninstall this test version first)

`flatpak install --user https://dl.flathub.org/build-repo/83617/com.github.taiko2k.tauonmb.flatpakref`

**Comment by iamalexei** [comment]
@Taiko2k still no luck

**Comment by iamalexei** [comment]
Ironically, VLC's and Audacious' PipeWire plugins are not able to change the sample rate of the driver--they only set it for the player itself, ergo making the 44.1 kHz sample rate passthrough unachievable.

**Comment by Taiko2k** [comment]
That's how it should work, apps should be able to send whatever format to the sound server and it's up to the sound server how it handles that. 

I think what miniaudio is doing is introducing resampling for "compatibility" reasons with pulseaudio. I thought a PR I sent fixed that but I can't remember the details.

**Comment by iamalexei** [comment]
> That's how it should work, apps should be able to send whatever format to the sound server and it's up to the sound server how it handles that.

When I play 44.1 audio with mpv (pipewire/pulse), pw-play, VLC (pulse) or Audacious (pulse), I get:

```
$ cat /proc/asound/card0/pcm0p/sub0/hw_params | grep rate
rate: 44100 (44100/1)
```

But the same track with Audacious (pipewire) or VLC (pipewire) shows:

```
$ cat /proc/asound/card0/pcm0p/sub0/hw_params | grep rate
rate: 48000 (48000/1)
```

**Comment by giostark** [comment]
Hi all ,
Arch user , latest software...
I tried to use only pulseaudio and removed entirely pipewire but no luck. Even tried to use just the alsa without pulse and pipewire. Audacious player switch between frequencies using any output plugin (pipewire , pulseaudio , alsa) . 
Unfortunately Tauon do not switch at all. In any possible combination. (avoid resampling enabled)
This is really frustrating because this player is one of the best around. 
My audio chain is good enough to let clearly ear the difference between a 44.1 file vs a 192 file (natively registered/converted at those  frequencies). 
Hopefully something will change in future... Anyway thanks a lot for your amazing work ...

**Comment by giostark** [comment]
Just an update...
Under Arch the funny part is that now this work exactly the opposite :-) (after updates)
The sample rate change always despite you tell it to stay fixed !!!
Hahaha ... 
Then , after an update wireplumber killed ALL the audio devices  (3 audio devices disappeared)   and I had to switch back to pipewire-media-session to have sound back again.
Fine , I'll disable the NOS function and I'll wait that both the possibilities will be available.

**Comment by Kevinhykuo** [comment]
The bug still exists in Tauon 10.0.1. Using PipeWire locks the output to 48kHz, even with the "Avoid Resampling" option turned on. The only solution (so far) to get the sampling rate to change dynamically with the audio track is to select PulseAudio with "Avoid Resampling" turned on. 
(My OS: Kubuntu 24.04 LTE)

**Comment by mirthquake** [comment]
Yup, I can confirm this bug still exists in Tauon 10.0.1 as well. Switching to PulseAudio with "Avoid Resampling" turned on does fix the sampling rate issue but introduces a whole lot of crackling for some reason, so it's not really a workable alternative.

<img width="1332" height="635" alt="Image" src="https://github.com/user-attachments/assets/a0c13eb0-1844-4c2b-863e-859ea2bbe694" />

Here are the results when playing a song sampled at 96 kHz on Tauon.

<img width="1304" height="457" alt="Image" src="https://github.com/user-attachments/assets/5f4f98f9-3f82-4de2-9e8f-d433a967e6fd" />

FWIW, playing it on Rhythmbox, Strawberry, and Quodlibet switches the sampling successfully to 96 kHz.

**Comment by giostark** [comment]
Hi all , I try to be supportive ...
The problem do not seem to be Tauon alone but also the sub system and how Tauon work with it.
If you want see change the default sound rate of Pulse or Pipewire on your DAC and in the console (pw-top) you have to enable this feature  manually in the dedicate config files .
If you do not enable this features in those files Tauon will no change the sound rate accordingly to the track you are listening. (as for its setting)
So , If you are using only Pipewire locate the config files : https://docs.pipewire.org/page_man_pipewire_conf_5.html
```
/etc/pipewire/pipewire.conf
/usr/share/pipewire/pipewire.conf
/usr/share/pipewire/pipewire.conf.d/
/etc/pipewire/pipewire.conf.d/
```
Then , search for the line : #default.clock.allowed-rates = [ 48000 ] 
And make it like this (but this depend also on what your DAC support)  : 
default.clock.allowed-rates = [ 44100,48000,88200,96000,176400,192000,352800,384000,512000] 

Important , if you have uncommented :  #default.clock.rate  = 48000 , you have to comment it. Together with allowed rate it will not work. 

If you are using only PulseAudio then you have to search for the config files as described here  https://wiki.archlinux.org/title/PulseAudio
Located the daemon.conf config file you should search for : alternate-sample-rate and add the same settings as above for the allowed clock.

I do use Arch and I succeeded this sample rate change on my DAC using Pipewire. Today lots of system do use Pipewire as default and to use only Pulse could bring issues of compatibility. In Arch you could have to load pipewire.service, pipewire-pulse.service and wireplumber.service user units if necessary.
So I suggest to understand properly how the single distribution do work and how it integrate the sound system. 
In this scenario is difficult for Tauon alone to create a combo of settings to make this changing rate effective on all distros. 
Maybe I was not entirely pertinent about the issue originally reported but I hope that some point can be useful to reach a working solution for each are facing this problem.

**Comment by Taiko2k** [comment]
Ive committed a fix

**Comment by giostark** [comment]
> FWIW, playing it on Rhythmbox, Strawberry, and Quodlibet switches the sampling successfully to 96 kHz.

Some player are able to direct connect to the hardware(if you select the right output in the settings of the player) jumping the Pipewire or Pulse Sound System. Those players have this ability for make the DSD format playable . The DSD ( Direct Stream Digital ) as for its nature bypass the Sound System and connect directly to the hardware.
At this point you have to regulate the audio volume from the DAC or the Amplifier and not from the PCM volume of the system.
I do have so few DSD and to be honest my system is good enough that this annoying feature is  not necessary at all.
Reading your posts remember me my journey to make this thing work :-)

**Comment by mirthquake** [comment]
Just upgraded to Tauon v11 and this issue appears to be fixed! Thanks a lot :-)
